proof/access-control/AARCH64/ArchAccess.thy

```diff
@@ -0,0 +1,363 @@
+  for aag subjects l ao ao' where
+  arch_troa_asidpool_clear:
+    "\<lbrakk> ao = ASIDPool pool; ao' = ASIDPool pool';
+       asid_pool_integrity subjects aag pool pool' \<rbrakk>
+       \<Longrightarrow> arch_integrity_obj_atomic aag subjects l ao ao'"
+
+inductive arch_integrity_obj_alt ::
+   "'a PAS \<Rightarrow> 'a set \<Rightarrow> 'a \<Rightarrow> arch_kernel_obj \<Rightarrow> arch_kernel_obj \<Rightarrow> bool"
+  for aag subjects l' ao ao' where
+  arch_tro_alt_asidpool_clear:
+    "\<lbrakk> ao = ASIDPool pool; ao' = ASIDPool pool';
+       asid_pool_integrity subjects aag pool pool'\<rbrakk>
+       \<Longrightarrow> arch_integrity_obj_alt aag subjects l' ao ao'"
+
+definition auth_ipc_buffers :: "'z::state_ext state \<Rightarrow> obj_ref \<Rightarrow> obj_ref set" where
+  "auth_ipc_buffers s \<equiv> \<lambda>p. case (get_tcb p s) of
+     None \<Rightarrow> {}
+   | Some tcb \<Rightarrow>
+     (case tcb_ipcframe tcb of
+        ArchObjectCap (FrameCap p' R vms False _) \<Rightarrow>
+          if AllowWrite \<in> R
+          then (ptr_range (p' + (tcb_ipc_buffer tcb && mask (pageBitsForSize vms))) msg_align_bits)
+          else {}
+      | _ \<Rightarrow> {})"
+
+end
+
+
+context begin interpretation Arch .
+
+requalify_consts
+  vspace_cap_rights_to_auth
+  aobj_ref'
+  acap_asid'
+  state_vrefs
+  state_asids_to_policy_arch
+  integrity_asids
+  ctxt_IP_update
+  arch_IP_update
+  arch_cap_auth_conferred
+  arch_integrity_obj_atomic
+  arch_integrity_obj_alt
+  auth_ipc_buffers
+
+requalify_facts
+  integrity_asids_updates
+  state_vrefs_upd
+  integrity_asids_kh_upds
+  integrity_asids_kh_upds'
+  integrity_asids_kh_update
+
+end
+
+declare state_vrefs_upd[simp]
+declare integrity_asids_updates[simp]
+
+
+context Arch begin global_naming AARCH64
+
+(* FIXME AARCH64: update the access control spec
+   -Parameterise arch object updates with machine state and arch state
+   -Model virtualised machine state more explicitly
+   -Specify when virtualised machine state can change (i.e. restoring from current VCPU)
+   -Specify when arm_current_vcpu can change (i.e. once current VCPU has been saved)
+   -Specify integrity constraints for FPUs
+*)
+
+\<comment> \<open>Anyone can save virtualised registers to the current VCPU\<close>
+lemma arch_troa_vcpu_save_reg:
+    "\<lbrakk> aobjs_of s vptr = Some (VCPU vcpu); ao' = VCPU vcpu';
+       option_map fst (arm_current_vcpu (arch_state s)) = Some vptr;
+       vcpu' = vcpu\<lparr>vcpu_regs := (vcpu_regs vcpu)(reg := vcpuHardwareReg_val reg (machine_state s))\<rparr> \<rbrakk>
+       \<Longrightarrow> arch_integrity_obj_atomic aag subjects l ao ao'"
+  sorry
+
+(* FIXME AARCH64: assert a connection to the current (or soon-to-be-switched-to) thread? *)
+\<comment> \<open>Anyone can update the virtual count offset in the current VCPU\<close>
+lemma arch_troa_vcpu_restore_vtimer:
```

Comment:
The offset update happens only after we have switched the current VCPU to be active, which implies that it is the VCPU of the current thread. This means we might be able to drop this rule, because the change is happening in the current label, which is allowed to change anything that belongs to it already anyway.

Reply:
Thinking more about it, the `integrity` predicate is going to refer to the current VCPU in the initial state, though, not in the final state. So maybe we do need this clause. Still something to keep an eye on for an opportunity to tighten up the possible state changes more.

Reply:
After more discussion offline, we probably do need this, so I'll mark it as resolved.